### ISSUE TITLE
Wrong values after simple query on mongodb since 6.2.1

### ISSUE THREAD

**Issue by nazarkulyk** [Automattic/mongoose#11526]
Hallo,

i found some regression in 6.2.1 version. It did worked all times till including 6.2.0 version.

MongoDB Data:
```js
{ _id: 60817c815a30530098dd0d6f
name : "Morgen Star New"
details: Object
  prospectorDetails: Object
    name: "Günter Lüschen 10"
  exhibitorDetails: Object
    name: "Günter Lüschen 20"
}
```

Output from mongoose 6.2.1 and higher:

```js
{ _id: new ObjectId("60817c815a30530098dd0d6f"),
  name: 'Günter Lüschen 20',
  details: {
    prospectorDetails: { name: 'Günter Lüschen 10' },
    exhibitorDetails: { name: 'Günter Lüschen 20' }
  }
}
```

Output from mongoose 6.2.0 and lower:

```js
{ _id: new ObjectId("60817c815a30530098dd0d6f"),
  name: 'Morgen Star New',
  details: {
    prospectorDetails: { name: 'Günter Lüschen 10' },
    exhibitorDetails: { name: 'Günter Lüschen 20' }
  }
}
```

Mongoose >= 6.2.1 breaks "name" attribute of root object.

Schema:
```typescript
export const HorseSchema = new Schema<Horse>(
  {
    name: { type: SchemaTypes.String, required: false, index: true }
  },
  { strict: false, strictQuery: false }
);
```

**Comment by IslandRhythms** [comment]
Modify this script to reproduce your bug.

```javascript
const mongoose = require('mongoose');

const testSchema = new mongoose.Schema({
    name: String,
    details: {
        person: {
            name: String
        }
    }
}, {strict: false, strictQuery: false});

const Test = mongoose.model('Test', testSchema);

async function run() {
    await mongoose.connect('mongodb://localhost:27017');
    await mongoose.connection.dropDatabase();

    await Test.create({
        name: 'Foo',
        details: {
            person: {
                name: 'Bar'
            }
        }
    });

    console.log(await Test.findOne())
}

run();
```

**Comment by jeanbmar** [comment]
@IslandRhythms here it is:

```
import '../dotenv.mjs';
import connect from '../lib/helpers/mongoose-connect.mjs';

const mongoose = await connect();

const testSchema = new mongoose.Schema(
  {
    name: String,
  },
  { strict: false }
);

const Test = mongoose.model('Test', testSchema);

async function run() {
  await mongoose.connect('mongodb://localhost:27017');
  await mongoose.connection.dropDatabase();

  await Test.create({
    name: 'Foo',
    details: {
      person: {
        name: 'Bar',
      },
    },
  });

  const test = await Test.findOne({}).exec();
  console.log(test.name); // prints Bar
}
run();
```

**Comment by nazarkulyk** [comment]
There is realy issue with mongoos >=2.6.1

There are two documents in collention. One returns wrong "name":

![image](https://user-images.githubusercontent.com/22763871/158361403-af943ee2-e1fb-4814-9dcc-573642590e91.png)

![image](https://user-images.githubusercontent.com/22763871/158361774-ae420707-fcc7-40f7-a8e2-5a4e85a905f1.png)

![image](https://user-images.githubusercontent.com/22763871/158362064-31894c28-3b83-431e-9b29-68d531160ed4.png)

The difference is "key" order on Mongo DB self in collection. If attribute is last one, what happends in your case @IslandRhythms then everything works as expected, but if in document attribute is before nested - it takes last one.
See screenshots

**Comment by nazarkulyk** [comment]
I definitively can reproduce it if i create object on Mongoo DB manualy. So the Object looks like:

```json
{"_id":{"$oid":"62309a639c0470bef6c6febd"},"name":"Full Name","person":{"firstName":{"name":"first Name"},"lastName":{"name":"last Name"}}}
```

After that:

```js
const test1 = await Test.findOne({}).exec();
console.log(test1);
const test2 = await Test.findOne({}).select('name').exec();
console.log(test2);
```

Output is:

```txt
❯ node src/test.js
Mongoose: tests.findOne({}, { projection: {} })
{
  _id: new ObjectId("62309a639c0470bef6c6febd"),
  name: 'last Name',
  person: {
    firstName: { name: 'first Name' },
    lastName: { name: 'last Name' }
  }
}
Mongoose: tests.findOne({}, { projection: { name: 1 } })
{ _id: new ObjectId("62309a639c0470bef6c6febd"), name: 'Full Name' }
```

**Comment by nazarkulyk** [comment]
Here is even better script to reproduce. There is not "name" filled on create. But we get it if query:

```js
const mongoose = require('mongoose');

const testSchema = new mongoose.Schema(
  {
    name: String
  },
  { strict: false, strictQuery: false }
);

const Test = mongoose.model('Test', testSchema);

async function run() {
  await mongoose.connect('mongodb://localhost:27017');
  await mongoose.connection.dropDatabase();

  await Test.create({
    details: {
      person: {
        name: 'Bar'
      }
    }
  });

  console.log(await Test.findOne());
}

run();
````

Result:

```text
❯ node src/test.js
{
  _id: new ObjectId("62309efc4de374c11812fbbe"),
  details: { person: { name: 'Bar' } },
  name: 'Bar',
  __v: 0
}
```

**Comment by IslandRhythms** [comment]
@jeanbmar Your script does not reproduce it.
@nazarkulyk Your script reproduces it.

```javascript
const mongoose = require('mongoose');

const testSchema = new mongoose.Schema(
  {
    name: String
  },
  { strict: false, strictQuery: false }
);

const Test = mongoose.model('Test', testSchema);

async function run() {
  await mongoose.connect('mongodb://localhost:27017');
  await mongoose.connection.dropDatabase();

  await Test.create({
    details: {
      person: {
        name: 'Bar'
      }
    }
  });

  console.log(await Test.findOne());
}

run();
```

**Comment by vkarpov15** [comment]
This issue was introduced when we fixed #11309 with https://github.com/Automattic/mongoose/commit/d7649f2dd77f02d9cad74956eb936d7deec3b003. Sorry for the trouble - the fix will be in v6.2.8.
